rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+
+    pub fn name(&self) -> String;
+
+    pub fn merge(
+        &self,
+        key: Bytes,
+        existing_value: Bytes,
+        value: Bytes
+    ) -> Result<Bytes, MergeOperatorError>;
+}
+```
+
+Initially, there will be a single optional instance of the merge operator per database. The user needs to ensure that the compactor and writer run with the same merge operator to get predictable results.
+
+## Implementation
+
+### Data Structures
+
+We need to introduce an additional variant of `ValueDeletable` that denotes a **Merge Operand**. As a result of this change, we can no longer represent the deletable as `Option<Bytes>` or `Option<&[u8]>`, because a non-empty state can represent either a value or a merge operand. This requires reusing the ValueDeletable data structure throughout the stack—all the way from the memtable to the row codec.
+
+```rust
+#[derive(Debug, Clone, PartialEq)]
+pub enum ValueDeletable {
+    ...
+    Merge(Bytes)
+    ...
+}
+```
+
+### Row Codec
+
+We need to slightly adjust the binary format of the SST file and introduce a new flag to denote a merge operand. Note that the `Tombstone` and `Merge` flags are mutually exclusive.
+
+Aside from this, `Merge` rows can follow the same structure as `Value` rows.
+
+This change is backwards compatible.
+
+### Read Path
+
+Merges are expected to be read using the standard `get` API.
+
+For each key, SlateDB maintains a history of operations, which is subject to compaction.
+
+Let's denote each operation as `OPi`. A key (`K`) that has experienced `n` changes looks like this logically (physically, the changes could be spread across the memtable and multiple SST files):
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+
+earliest --------------------> latest
+```
+
+Currently, during a read operation, SlateDB always observes the latest known state of the database. In other words, it only needs to look up the latest operation. If it's a `Value`, it simply returns it, and if it's a `Tombstone`, it returns `None`. All previous operations can be ignored and will be removed by compaction at some point.
+
+```
+K:    OP1    OP2    OP3    ...    OPn
+                                   ^
+                                   |
+                                  GET
+```
+
+With the addition of a **Merge Operand**, it may be necessary to look backwards at previous values, up to the point where we encounter either a `Value` or a `Tombstone`. Everything beyond that point can be ignored.
+
+```
+K:    OP1    OP2    OP3    OP4    OPn
+            Value  Merge  Merge  Merge
+                                   ^
+                                   |
+                                  GET
+```
+
+
+In the above example, `get` should essentially return something like:
+
+```
+Merge(OP2, Merge(OP3, Merge(OP4, OPn)))
+```
+
+### Compactions
+
+Compactions will continuously reduce the history by using the same mechanics as described above, performing partial merges along the way.
```

Comment:
Compaction has a set of source SSTs, and a destination SSTs. Those are the only SSTs that are read. Is compaction expected to reduce all merge operations to a single value, or is it expected to combine multiple merges into either a single value or a single merge operation? 

Compaction may not include the SST that contains the value.

Reply:
It needs to follow the same mechanism I've already described in a comment above.

```
(previous + new) => (result)
tombstone + value => value
tombstone + merge => value
merge + tombstone => tombstone
merge + merge => merge
value + merge => value
merge + value => value # (merge will be ignored)
```

I'm expanding the section to cover this. It needs to become an invariant otherwise we break correctness.